[4.1]Ease the get of inputs from dependant fields in api's

pxpm opened the pull request:
I found this issue while debugging relationships and dependencies as part of my work on #3065 

As it is now when you are using dependencies with ajax fields, the docs says the way to grab the inputs is to use: 

```php
$form = collect($request->input('form'))->pluck('value', 'name'); 
```

If your form have any select multiple field you would endup with `articles[]` or `categories[]` in your form request, as many times as many of those entities you have selected. When you grab the `request('form')` this is what you get:

![input_form](https://user-images.githubusercontent.com/7188159/90025933-9aef0f80-dcae-11ea-9498-c1ef92ca5031.png)

After you call `pluck('value', 'name')` on it you get:

![collect](https://user-images.githubusercontent.com/7188159/90025990-ab9f8580-dcae-11ea-8e78-691d50a35d6d.png)

**### Notice we only have one of the selected `articles[]` and one of the selected `categories[]` ?** 

Also the names of the `HasOne` relations comes as they are in html `address[street]` but developer used `address.street` as the field name. 

So with this PR the result of `$form_values = backpack_input_parse(request('form'));` would be:

![Screenshot_1](https://user-images.githubusercontent.com/7188159/90026427-2ff20880-dcaf-11ea-8d2d-4bf7037d7fb0.png)


So the docs should be updated accordingly to:

```php
    public function fetchCategory()
    {
        return $this->fetch([
            'model' =>  \Backpack\NewsCRUD\app\Models\Category::class,
            'query' => function($model) {
               
                $form = backpack_input_parse(request('form'));

                return $model->where('category_id', $form['category_id']);
            }
        ]);
    }

//same for api request
$form = backpack_input_parse(request('form'));
if(!is_null($form['field'])) {
}

```

Best,
Pedro

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/9a37bf0a-4d4b-42ba-9064-e406ae3ea3eb) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/9a37bf0a-4d4b-42ba-9064-e406ae3ea3eb) completed: **1 updated code elements**

promatik commented:
@pxpm I think this is a great helper 👌
Quick question on this, in your example, shouldn't address be an array too?

```php
'address' => [
    'street' => 'my new address',
    'icon_id' => 31,
]
```

Developers would access like `$form['address']['street']`, instead of `$form['address.street']` even though they defined the field as `address.street`.

promatik commented on src/helpers.php:
```diff
@@ -30,6 +30,45 @@ function backpack_authentication_column()
+     * @return array
+     */
+    function backpack_input_parse($input)
+    {
```
```suggestion
    {
        $input = $input ?? request('form');
```

Can/Should we ease this and assume most of the times developers are going to use `request('form')`?

pxpm replied:
I think we could, but in that case we should rename the function to `backpack_form_input_parse()` and assume that the parsed input is always a form request. 

Do you foresee any scenario where we could use this to parse other inputs that are not in `request('form')` ? Otherwise we will endup `request()->merge('form',$data)` to call `backpack_form_input_parse()` and have it in the request.

Let me know,
Pedro

promatik replied:
_(if we are keeping the helper function)_
I agree with the name change, and maybe we can go even further 😅 `backpack_form_input()` parse is implicit.
If it's parsed or not, developers just want to know they are obtaining the form input data ready to use.

---

> Do you foresee any scenario where we could use this to parse other inputs that are not in request('form') ? Otherwise we will endup request()->merge('form',$data) to call backpack_form_input_parse() and have it in the request.

So, firstly yes, I can't see how will anyone need this using anything other than the `request('form')`.
Do you suggest to merge/replace the data on the request?
Can/Should we do that by default?

pxpm commented:
I agree with array instead of dot. 

will refactor accordingly. Thanks @promatik

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/878aaa6b-4a58-4cea-ab20-f03aa20677ff) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/7fbe35a3-e0cb-4dd0-ad23-81eb8cd49be8) was created.

pxpm commented:
Just refactored according to our conversation @promatik , renamed to `backpack_form_input()`, removed the required parameter and refactored to return arrays instead of dot notation names.

I searched for a regex vs str functions and it seems str functions perform better than regex.

Let me know.

Best,
Pedro

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/878aaa6b-4a58-4cea-ab20-f03aa20677ff) completed: **1 new issues, 1 updated code elements**

pxpm commented on src/helpers.php:
```diff
@@ -30,6 +30,51 @@ function backpack_authentication_column()
+     */
+    function backpack_form_input()
+    {
+        $input = request('form');
```
```suggestion
        $input = request('form') ?? [];
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/0c021a6a-972d-4e56-b754-1c7d2b7512c7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/7fbe35a3-e0cb-4dd0-ad23-81eb8cd49be8) completed: **1 new issues, 1 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/0c021a6a-972d-4e56-b754-1c7d2b7512c7) completed: **1 new issues, 1 updated code elements**

promatik commented on src/helpers.php:
```diff
@@ -30,6 +30,51 @@ function backpack_authentication_column()
+            }
+
+            if (is_null($input_arg)) {
+                if (! isset($result[$input_key])) {
+                    $result[$input_key] = $row['value'];
+                } else {
+                    $result[$input_key] = ! is_array($result[$input_key]) ? [$result[$input_key]] : $result[$input_key];
+
+                    array_push($result[$input_key], $row['value']);
+                }
```
```suggestion
                if (!isset($result[$input_key])) {
                    $result[$input_key] = $start ? [$row['value']] : $row['value'];
                } else {
                    array_push($result[$input_key], $row['value']);
                }
```

@pxpm this fixes cases where you're only sending one entry.
```php
articles[] => 6
```
was becoming
```php
articles => 6
```
instead of
```php
articles => [6]
```

promatik commented:
@pxpm can you also add the PR to fix the docs?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/78dedabd-6b3e-4c37-b8fe-df78b61a1dc9) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/bacf83a2-39df-4768-a44b-bab81fe95ba0) was created.

pxpm commented:
Docs here: https://github.com/Laravel-Backpack/docs/pull/224

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/bacf83a2-39df-4768-a44b-bab81fe95ba0) completed: **1 new issues, 1 updated code elements**
